src/cocotb/handle.py

```diff
@@ -573,142 +572,317 @@ class _GPISetAction(enum.IntEnum):
+else:
+
+    def _schedule_write(
+        handle: "ValueObjectBase[Any, Any]",
+        write_func: Callable[[int, ValueT], None],
+        action: _GPISetAction,
+        value: ValueT,
+    ) -> None:
+        if cocotb.sim_phase == cocotb.SimPhase.READ_WRITE:
+            # If we are already in the ReadWrite phase though, do it immediately as an optimization.
+            write_func(action.value, value)
+        elif action == _GPISetAction.DEPOSIT:
+            # Queue write for the beginning of the next ReadWrite phase because we can't trust the simulator. =(
+            if handle in _write_calls:
+                del _write_calls[handle]
+            _write_calls[handle] = (write_func, action, value)
+            _writes_pending.set()
+        else:
+            # If we are writing anything that isn't an inertial write, it must be applied immediately.
+            write_func(action.value, value)
+
+
+#: Type returned by the :attr:`~ValueObjectBase.value` getter and returned by the :meth:`~ValueObjectBase.get` method.
+ValueGetT = TypeVar("ValueGetT")
+
+
+#: Type accepted by the :attr:`~ValueObjectBase.value` setter and the :meth:`~ValueObjectBase.set` and :meth:`~ValueObjectBase.setimmediatevalue` methods.
 ValueSetT = TypeVar("ValueSetT")
 
 
-class ValueObjectBase(SimHandleBase, Generic[ValuePropertyT, ValueSetT]):
-    """Base class for all simulation objects that have a value."""
+class ValueObjectBase(SimHandleBase, Generic[ValueGetT, ValueSetT]):
+    """Abstract base class for simulation objects that have value."""
 
     @property
-    @abstractmethod
-    def value(self) -> ValuePropertyT:
+    def value(self) -> ValueGetT:
         """Get or set the value of the simulation object.
 
-        :getter: Returns the current value of the simulation object.
+        :getter: Return the current value of the simulation object.
 
         :setter:
-            Assigns the value at end of the current simulator delta cycle.
-            Takes whatever values that :meth:`set` takes,
-            including :class:`Deposit`, :class:`Force`, :class:`Freeze`, and :class:`Release` actions.
-
-        .. note::
-
-            Use :meth:`setimmediatevalue` if you need to set the value of the simulation object immediately.
+            Set the value of the simulation object.
+
+            See :class:`Deposit`, :class:`Force`, :class:`Freeze`, :class:`Release`, and :class:`Immediate`
+            for additional actions that can be taken when setting a value.
+            These are used like so:
+
+            .. code-block:: python
+                dut.handle.value = 1  # default Deposit action
+                dut.handle.value = Deposit(2)
+                dut.handle.value = Force(3)
+                dut.handle.value = Freeze()
+                dut.handle.value = Release()
+                dut.handle.value = Immediate(4)
+
+            .. note::
+                The Python type system thinks the setter only supports the getter's type,
```

Comment:
Recently changed: https://github.com/python/mypy/pull/18510

Reply:
Interesting... I'm not sure I agree with that actually, but okay. I'll remove this note.